### PR TITLE
Make `ArrayNode.elements()` return `ListIterator()`

### LINKED ISSUE #3120: Return `ListIterator` from `ArrayNode.elements()`
**Problem/Requirement:**
I have the requirement to traverse over Json document models and do selective manipulation on specific content by replacing some ValueNodes by other ValueNodes.

So the approach is, to traverse over the ObjectNode/ArrayNodes using the iterators provided via fields() (ObjectNode) respective elements() (ArrayNode).

Since for the ObjectNode the iterated object is the Map.Entry interface I can easily set a different JsonNode via entry.setValue(JsonNode value) without considering context:

```Java
Iterator<Map.Entry<String, JsonNode>> fields = currentNode.fields();
while (fields.hasNext()) {
  Map.Entry<String, JsonNode> entry = fields.next();
  entry.setValue( deriveNode(entry.getValue()) );
}
```

However, for ArrayNode, there is no approach to selectively replace single JsonNodes while iterating over the content.
To replace single fields, currently I need to:
1. Store the Parent ArrayNode
2. track the index I am working on
3. Call a parent.set(index, JsonNode)

Worse: the modification invalidates the iterator. Subsequent iteration calls will cause a ConcurrentModificationException.

Thus, after that, I need to re-obtain a new iterator instance and iterate index-n times to return to the previous position in order to continue from where I started.

**Proposed Solution**
A possible solution would be to internally use the ListIterator instead of an Iterator when calling ArrayNode.elements().
With a ListIterator, it is possible to cast the iterator to ListIterator and then call iterator.set(JsonNode value):

In ArrayNode.java:
```Java
@Override
public Iterator<JsonNode> elements() {
  return _children.listIterator();
 }
```

I do understand that changing the officially released method signature for the ArrayNode.elements() might cause compatibility issues therefore I do not request this.
Only under-the-hood-is the internally used instance could be of type listIterator, that would suffice. The user then has the potential to cast to ListIterator, since he has control over the specific version of the library used.

**Usage example**
With this small modification, the following can easily be done:
```Java
ListIterator<JsonNode> elements = (ListIterator<JsonNode>)currentNode.elements();
while (elements.hasNext()) {
  JsonNode childNode = elements.next();
  elements.set(deriveNode(childNode)); 
}
```

### PR DESCRIPTION
**Pull request by JooHyukKim**
Resolve #3120

### CODE REVIEW

File: src/main/java/com/fasterxml/jackson/databind/node/ArrayNode.java [hunk 1]
```diff
@@ -248,7 +248,7 @@ public int size() {
 
     @Override
     public Iterator<JsonNode> elements() {
-        return _children.iterator();
+        return _children.listIterator();
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/node/ArrayNode.java:251]: Could you add a comment wrt references to issue, and version in which change goes in (2.18)?

**Review by cowtowncoder — APPROVED** [review]
approved, would like comment & release notes entry, then can go in 2.18(.0).

**EDIT*: will make changes, merge.

File: src/main/java/com/fasterxml/jackson/databind/node/ArrayNode.java (hunk 1 above)
Comment (cowtowncoder) [reply src/main/java/com/fasterxml/jackson/databind/node/ArrayNode.java:251]: Actually n/m I can do this.

**Comment by JooHyukKim** [comment]
> approved, would like comment & release notes entry, then can go in 2.18(.0).
> 
> *_EDIT_: will make changes, merge.

I forgot about it. Thanks! 👍🏼

**Comment by cowtowncoder** [comment]
@JooHyukKim np! Thanks for the PR.

**Pull request merged**